python/hidet/distributed/partition/rule.py

```diff
@@ -0,0 +1,256 @@
+            if len(new_inputs) < len(inputs):
+                continue
+            outputs = op.reforward(new_inputs)
+        except (ValueError, RuntimeError, AssertionError, IndexError):
+            continue
+
+        # find the sharded output dimension
+        # for each output tensor, there should be at most one dimension being sharded
+        out_shard_dims = []
+        for i, out in enumerate(outputs):
+            shard_dim = -1
+            origin_shape = op.outputs[i].shape
+            for dim in range(len(out.shape)):
+                if out.shape[dim] == origin_shape[dim] // num_shards:
+                    if shard_dim != -1:
+                        shard_dim = None  # invalid output shape
+                        # Because there should be at most one sharded dimension
+                    else:
+                        shard_dim = dim
+                elif out.shape[dim] != origin_shape[dim]:
+                    shard_dim = None  # invalid output shape
+            if shard_dim is not None:  # output is invalid result of a sharding
+                out_shard_dims.append(shard_dim)
+            else:
+                break
+
+        if len(out_shard_dims) == len(outputs):  # shape analysis valid
+            # data dependency analysis
+            in_specs = [TensorShardSpec(len(i.shape), shard_dim) for i, shard_dim in zip(inputs, in_shard_dims)]
+            out_specs = [TensorShardSpec(len(o.shape), shard_dim) for o, shard_dim in zip(outputs, out_shard_dims)]
+            if data_dependency_analyzer.check(in_shard_dims, out_shard_dims):
+                found_rules.append(OpShardSpec(in_specs, out_specs))
+            else:  # Failed, but it still can be partial results. Try to fix it with reduction.
+                if data_dependency_analyzer.check(in_shard_dims, out_shard_dims, shard_reduce=True):
+                    reduce_op = op.task.outputs[0].value.reduce_operation
+                    found_rules.append(OpShardSpec(in_specs, out_specs, reduce_fn=[AllReduce(reduce_op)]))
+                    # if all_reduce is valid, then reduce_scatter is also valid
+                    output = op.task.outputs[0]
+                    if output.shape[0] % num_shards == 0:
+                        out_spec = TensorShardSpec(len(output.shape), 0)
+                        found_rules.append(OpShardSpec(in_specs, [out_spec], reduce_fn=[ReduceScatter(reduce_op)]))
+
+    return found_rules
+
+
+if __name__ == '__main__':
+    import hidet.testing
+
+    def resnet_test():
+        model = hidet.testing.models.resnet.resnet18()
+        x = hidet.symbol([8, 3, 224, 224])
+        flow_graph = hidet.trace_from(model(x))
+
+        cache = {}
+        for op in flow_graph.nodes:
+            print(op)
+            if str(op) not in cache:
+                shard_plans = op_shard_rule_search(op, 4)
+                cache[str(op)] = shard_plans
+            for sp in cache[str(op)]:
+                print(sp)
+
+    resnet_test()
```

Comment:
Add some tests for the shard rule searching in `tests/distributed/partition/test_shard_rule.py`.